staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go

```diff
@@ -225,11 +229,15 @@ func RunOrDie(ctx context.Context, lec LeaderElectionConfig) {
 // GetLeader returns the identity of the last observed leader or returns the empty string if
 // no leader has yet been observed.
 func (le *LeaderElector) GetLeader() string {
+	le.mutex.RLock()
+	defer le.mutex.RUnlock()
 	return le.observedRecord.HolderIdentity
 }
 
 // IsLeader returns true if the last observed leader was this client else returns false.
 func (le *LeaderElector) IsLeader() bool {
+	le.mutex.RLock()
+	defer le.mutex.RUnlock()
```

Comment:
with the lock, the data race can still happen, since the leadership may change after the function is called. I'm not sure what's the use case of allowing a second caller other than the leader election code itself to call `IsLeader()`

Reply:
I’m using leaderelection package for distributed locking of API applications.
(I know this is not the use case that was originally envisioned.)

Create a lock for each API request scope to prevent multiple containers or goroutines from manipulating the same resource at the same time.
Although it may not be the original usage, I personally find it very helpful because I can implement distributed locking using Kubernetes API without using Redis or other databases.

Currently, because of the data race, we do not use the `IsLeader()` function and use the Kubernetes API to get the lease object, but it will be simpler to implement once the data race is fixed.

example code:

<details>

```go
package leaderelection

import (
	"context"
	"time"

	coordinationv1 "k8s.io/api/coordination/v1"
	"k8s.io/apimachinery/pkg/api/errors"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/apimachinery/pkg/util/wait"
	"k8s.io/client-go/kubernetes"
	"k8s.io/client-go/rest"
	"k8s.io/client-go/tools/leaderelection"
	"k8s.io/client-go/tools/leaderelection/resourcelock"
	"sigs.k8s.io/controller-runtime/pkg/client"
)

const leaseLockNamespace = "foo-namespace"

type Client struct {
	*leaderelection.LeaderElector

	k8sClient client.Client
	lockName  string
	identity  string
}

func NewClient(config *rest.Config, k8sClient client.Client, lockName string, identity string) (*Client, error) {
	client, err := kubernetes.NewForConfig(config)
	if err != nil {
		return nil, err
	}

	lock := &resourcelock.LeaseLock{
		LeaseMeta: metav1.ObjectMeta{
			Name:      lockName,
			Namespace: leaseLockNamespace,
		},
		Client: client.CoordinationV1(),
		LockConfig: resourcelock.ResourceLockConfig{
			Identity: identity,
		},
	}

	leaderElectionConf := leaderelection.LeaderElectionConfig{
		Lock: lock,
		// release the lock by canceling the context
		ReleaseOnCancel: true,
		LeaseDuration:   60 * time.Second,
		RenewDeadline:   15 * time.Second,
		RetryPeriod:     5 * time.Second,
		Callbacks: leaderelection.LeaderCallbacks{
			OnStartedLeading: func(_ context.Context) {},
			OnStoppedLeading: func() {},
		},
	}

	elector, err := leaderelection.NewLeaderElector(leaderElectionConf)
	if err != nil {
		return nil, err
	}

	return &Client{
		LeaderElector: elector,
		k8sClient:     k8sClient,
		identity:      identity,
		lockName:      lockName,
	}, nil
}

// maxBackoffInterval defines the maximum amount of time to wait between
// attempts to become the leader.
const maxBackoffInterval = time.Second * 16

// Become will run a leader election and wait until it gets the leader.
// Cancel context to release the lock.
func (l *Client) Become(ctx context.Context) error {
	go l.Run(ctx)

	backoff := time.Second

	for {
		var lease coordinationv1.Lease

		select {
		case <-ctx.Done():
			return ctx.Err()
		default:
			err := l.k8sClient.Get(ctx, client.ObjectKey{Name: l.lockName, Namespace: leaseLockNamespace}, &lease)
			if !errors.IsNotFound(err) && err != nil {
				return err
			}
		}

		if l.isLeader(lease) {
			return nil
		}

		select {
		case <-time.After(wait.Jitter(backoff, .2)):
			if backoff < maxBackoffInterval {
				backoff *= 2
			}
		case <-ctx.Done():
			return ctx.Err()
		}
	}
}

func (l *Client) isLeader(lease coordinationv1.Lease) bool {
	if lease.Spec.HolderIdentity == nil {
		return false
	}

	return l.identity == *lease.Spec.HolderIdentity
}
```

</details>

Reply:
That's an interesting use case. Thanks for sharing.